packages/js/product-editor/src/components/details-categories-field/use-category-search.ts

```diff
@@ -272,9 +278,33 @@ export const useCategorySearch = () => {
 		[ categoriesAndNewItem ]
 	);
 
+	/**
+	 * this is the same as getFilteredItems but for the SelectTree component, where item id is a string.
+	 * After all the occurrences of getFilteredItems are migrated to use SelectTree,
+	 * this can become the standard version
+	 */
+	const getFilteredItemsForSelectTree = useCallback(
+		(
+			allItems: TreeItemType[],
+			inputValue: string,
+			selectedItems: TreeItemType[]
+		) => {
+			const searchRegex = new RegExp( escapeRegExp( inputValue ), 'i' );
+			return allItems.filter(
+				( item ) =>
+					selectedItems.indexOf( item ) < 0 &&
+					( searchRegex.test( item.name ) ||
+						( categoryTreeKeyValues[ +item.id ] &&
+							categoryTreeKeyValues[ +item.id ].isOpen ) )
+			);
+		},
+		[ categoriesAndNewItem ]
+	);
+
 	return {
 		searchCategories,
 		getFilteredItems,
+		getFilteredItemsForSelectTree,
 		categoriesSelectList: categoriesAndNewItem[ 0 ],
 		categories: categoriesAndNewItem[ 1 ],
 		isSearching,
```

Comment:
A lot of this logic can be removed once we also replace the `TreeSelect` in the add new modal, as the `TreeControl` will handle a lot of this for us.